source/sigv4.c

```diff
@@ -794,6 +918,308 @@ static SigV4Status_t getCredentialScope( SigV4Parameters_t * pSigV4Params,
 
 /*-----------------------------------------------------------*/
 
+    static bool isTrimmableSpace( const char * value,
+                                  size_t index,
+                                  size_t valLen,
+                                  size_t trimmedLength )
+    {
+        bool ret = false;
+
+        assert( ( value != NULL ) && ( index < valLen ) );
+
+        /* Only trim spaces. */
+        if( isspace( value[ index ] ) )
+        {
+            /* The last character is a trailing space. */
+            if( ( index + 1 ) == valLen )
+            {
+                ret = true;
+            }
+            /* Trim if the next character is also a space. */
+            else if( isspace( value[ index + 1 ] ) )
+            {
+                ret = true;
+            }
+            /* It is a leading space if no characters have been written yet. */
+            else if( trimmedLength == 0U )
+            {
+                ret = true;
+            }
+        }
+
+        return ret;
+    }
+
+/*-----------------------------------------------------------*/
+
+    SigV4Status_t copyHeaderStringToCanonicalBuffer( const char * pData,
+                                                     size_t dataLen,
+                                                     uint32_t flags,
+                                                     char separator,
+                                                     CanonicalContext_t * canonicalRequest )
+    {
+        SigV4Status_t status = SigV4Success;
+        size_t index = 0;
+        size_t numOfBytesCopied = 0;
+        size_t buffRemaining;
+        char * pCurrBufLoc;
+
+        assert( ( pData != NULL ) && ( dataLen > 0 ) );
+        assert( canonicalRequest != NULL );
+        assert( canonicalRequest->pBufCur != NULL );
+
+        buffRemaining = canonicalRequest->bufRemaining;
+        pCurrBufLoc = canonicalRequest->pBufCur;
+
+        for( index = 0; index < dataLen; index++ )
+        {
+            /* If the header field is not in canonical form already, we need to check
+             * whether this character represents a trimmable space. */
+            if( !( flags & SIGV4_HTTP_HEADERS_ARE_CANONICAL_FLAG ) &&
+                isTrimmableSpace( pData, index, dataLen, numOfBytesCopied ) )
+            {
+                /* Cannot copy trimmable space into canonical request buffer. */
+            }
+            /* Remaining buffer space should at least accommodate the character to copy and the trailing ":" */
+            else if( buffRemaining <= 1 )
+            {
+                status = SigV4InsufficientMemory;
+                break;
+            }
+            else
+            {
+                /* Lowercase header key only. '\n' character marks the ending of value. */
+                if( separator == '\n' )
+                {
+                    *pCurrBufLoc = ( pData[ index ] );
+                }
```

Comment:
What is the reason for not using `tolower` for `\n` as the separator character? The comment does not give me a clear reasoning for the special handling

Reply:
When we are adding the value the separator will be '\n' and we don't want to lowercase the value.